### PR TITLE
bspline module and C backend speed-ups

### PR DESCRIPTION
**Pull request by kbwestfall**
This PR introduces some C code primarily to improve the bspline fitting performance.  Primary changes are:

- Moved pypeit.core.pydl.bspline into its own module, `pypeit.bspline`
- Introduced C backend functions to speed up bspline fitting
    - now require `extension_helpers` package to build pypeit and
      necessary files/code in `setup.py` to build the C code
    - C functions will be used by default, but code will revert to pure
      python, if there's some problem importing the C module
    - Added tests and pre-cooked data to ensure identical behavior
      between the pure python and C functions.
- Moved some basis function builders to pypeit.core.basis

Tests and dev-suite running.

### CODE REVIEW

File: pypeit/bspline/bspline.py [hunk 1]
```diff
@@ -0,0 +1,684 @@
...
+from IPython import embed
+
+import numpy as np
+
+from pypeit.core import basis
+
+try:
+    from pypeit.bspline.utilc import cholesky_band, cholesky_solve, solution_arrays, intrv, \
+                                     bspline_model
+except:
```
[the remark this answers is not in the document]
Comment (kbwestfall) [reply pypeit/bspline/bspline.py]: I tried that first, but the exception was something different.  So I figured if the import fails for any reason, that it should fall back to the python versions.

File: pypeit/bspline/bspline.py [hunk 2]
```diff
@@ -0,0 +1,684 @@
...
+#                                    bspline_model
+
+# TODO: Types are important for the C extension. Types should be
+# limited to int, float, bool!
+# TODO: May need to add hooks to utilc.py that do the type conversion,
+# but that should be a last resort for stability.
+# TODO: This whole module needs to be cleaned up.
+
+
+class bspline(object):
```
[the remark this answers is not in the document]
Comment (kbwestfall) [reply pypeit/bspline/bspline.py]: Noted, and thanks for the heads up.

File: pypeit/utils.py [hunk 3]
```diff
@@ -604,62 +606,91 @@ def bspline_fit(x,y,order=3,knots=None,everyn=20,xmin=None,xmax=None,w=None,bksp
...
+    lower : :obj:`int`, :obj:`float`, optional
+        Lower rejection threshold in units of sigma, defaults to 5
+        sigma.
+    maxiter : :obj:`int`, optional
+        Maximum number of rejection iterations, default 10. Set this
+        to zero to disable rejection.
+    nord : :obj:`int`, optional
         Order of B-spline fit
-    bkpt : :class:`numpy.ndarray`
+    bkpt : `numpy.ndarray`_, optinoal
```
[the remark this answers is not in the document]
Comment (kbwestfall) [reply pypeit/utils.py]: fixed

**Comment by kbwestfall** [comment]
Tests pass.  Cooked v0.13.5 uploaded.  Restarting devsuite.

```
test_arc.py .                                                            [  0%]
test_arcimage.py ..                                                      [  1%]
test_biasframe.py ..                                                     [  2%]
test_bitmask.py ....                                                     [  4%]
test_bpmimage.py ..                                                      [  5%]
test_bspline.py .........                                                [  9%]
test_buildimage.py .                                                     [  9%]
test_calibrations.py .......                                             [ 12%]
test_coadd.py .                                                          [ 13%]
test_cooked.py .                                                         [ 13%]
test_coremeta.py .                                                       [ 14%]
test_datacontainer.py ........                                           [ 17%]
test_detector.py ...                                                     [ 19%]
test_edgetrace.py .....                                                  [ 21%]
test_flatfield.py ..                                                     [ 22%]
test_flexure.py .                                                        [ 22%]
test_flux.py ....                                                        [ 24%]
test_fluxspec.py ..                                                      [ 25%]
test_load.py .                                                           [ 26%]
test_load_images.py ...........                                          [ 31%]
test_masterframe.py .                                                    [ 31%]
test_match.py ....                                                       [ 33%]
test_metadata.py .....                                                   [ 35%]
test_moment.py ......                                                    [ 38%]
test_msgs.py ..                                                          [ 39%]
test_opticalmodel.py ....                                                [ 41%]
test_parse.py ...                                                        [ 42%]
test_pca.py ....                                                         [ 44%]
test_processimage.py ..                                                  [ 45%]
test_procimg.py .                                                        [ 45%]
test_pydl.py .                                                           [ 46%]
test_pypeit.py ..                                                        [ 47%]
test_pypeitimage.py .                                                    [ 47%]
test_pypeitpar.py ................................                       [ 62%]
test_pypeitsetup.py .......                                              [ 65%]
test_qa.py .                                                             [ 65%]
test_save.py .                                                           [ 66%]
test_scienceimage.py ...                                                 [ 67%]
test_scripts.py ............                                             [ 72%]
test_setups.py ...............                                           [ 78%]
test_slitmask.py .                                                       [ 79%]
test_slittrace.py ...                                                    [ 80%]
test_specobj.py ....                                                     [ 82%]
test_specobjs.py ....                                                    [ 84%]
test_spectrographs.py ...................                                [ 93%]
test_traceimage.py .                                                     [ 93%]
test_utils.py ....                                                       [ 95%]
test_vcorr.py ..                                                         [ 96%]
test_wavecalib.py .                                                      [ 96%]
test_waveimage.py ..                                                     [ 97%]
test_wavemodel.py .                                                      [ 98%]
test_wavetilts.py ....                                                   [100%]

========================= 219 passed in 720.12 seconds =========================
```

**Comment by kbwestfall** [comment]
Okay, I added a parameter that dictates how saturated slits should be treated in the flat-field modeling.  Default is to crash.  Other options are to mask the slit or to continue.  See pypeitpar.py and flatfield.py.

Tests still pass.

I've restarted the dev suite.  Note that default `saturated_slits = crash` causes two of the devsuite datasets to fault:
```
keck_lris_red/multi_1200_9000_d680_1x2
keck_deimos/830G_M_8100_26
```
For these two, I've changed the pypeit files in the devsuite to have:
```
[calibrations]
    [[flatfield]]
        saturated_slits = continue
```

File: pypeit/flatfield.py [hunk 4]
```diff
@@ -520,14 +525,43 @@ def fit(self, debug=False):
...
+            # slit.
             good_frac = np.sum(onslit & (rawflat < nonlinear_counts))/np.sum(onslit)
+            # TODO: set the threshold to a parameter?
             if good_frac < 0.5:
-                # TODO: Used slit ID in these print statments instead of slit index
-                msgs.warn('Only {:4.2f}'.format(100*good_frac)
-                          + '% of the pixels on this slit are not saturated.' + msgs.newline()
-                          + 'Consider raising nonlinear_counts={:5.3f}'.format(nonlinear_counts) +
-                          msgs.newline() + 'Not attempting to flat field slit {:d}'.format(slit))
+                if tweak_slits:
```
Comment (jhennawi) [inline pypeit/flatfield.py]: Might be clearer to change the error message from  "Only ... not saturated" to "More than ... are saturated."

File: pypeit/flatfield.py [hunk 5]
```diff
@@ -680,8 +714,11 @@ def fit(self, debug=False):
             # 1/10th of a pixel, but do not allow a bsp smaller than
             # the typical sampling. Use the bspline class to determine
             # the breakpoints:
-            spat_bspl = pydl.bspline(spat_coo_data, nord=4,
-                                     bkspace=np.fmax(1.0/median_slit_width[slit]/10.0,
+#            spat_bspl = pydl.bspline(spat_coo_data, nord=4,
```
Comment (jhennawi) [inline pypeit/flatfield.py]: Remove.

File: pypeit/flatfield.py [hunk 6]
```diff
@@ -794,7 +831,8 @@ def fit(self, debug=False):
             twod_ivar_data = twod_gpm_data.astype(float)/(twod_sig**2)
             twod_sigrej = 4.0
 
-            poly_basis = pydl.fpoly(2.0*twod_spat_coo_data - 1.0, npoly).T
+#            poly_basis = pydl.fpoly(2.0*twod_spat_coo_data - 1.0, npoly).T
```
Comment (jhennawi) [inline pypeit/flatfield.py]: Remove.

File: pypeit/reduce.py [hunk 7]
```diff
@@ -172,19 +172,17 @@ def extract(self, global_sky, sobjs_obj):
...
-                            self.sobjs.ech_objid > 0)  # pos indices of objects for this slit
-                    sobj = self.sobjs[np.where(thisobj)[0][0]]
-                    # Plate scale
-                    plate_scale = self.spectrograph.order_platescale(sobj.ECH_ORDER,
-                                                                 binning=self.binning)[0]
-                else:
-                    sobj = self.sobjs[iord]
-                    plate_scale = self.spectrograph.detector[self.det - 1]['platescale']
-                '''
+#                if self.pypeline == 'Echelle':
```
Comment (jhennawi) [inline pypeit/reduce.py]: Looks like this logic was moved to the get_platescale function so we can delete this commented out text.

File: pypeit/spectrographs/keck_hires.py [hunk 8]
```diff
@@ -219,55 +219,53 @@ def __init__(self):
...
-                        saturation      = 65535.,
-                        nonlinear       = 0.86,
-                        numamplifiers   = 1,
-                        gain            = 0.84, # high gain, low gain 2.2
-                        ronoise         = 3.1,
-                        suffix          = '_03'
-                        ),
-        ]
-        '''
+#        self.detector = [
```
Comment (jhennawi) [inline pypeit/spectrographs/keck_hires.py]: Delete?

File: pypeit/tests/test_bspline.py [hunk 9]
```diff
@@ -0,0 +1,201 @@
+
+import time
+import pytest
+
+from IPython import embed
```
Comment (jhennawi) [inline pypeit/tests/test_bspline.py]: Nice!

File: pypeit/utils.py [hunk 10]
```diff
@@ -604,62 +606,91 @@ def bspline_fit(x,y,order=3,knots=None,everyn=20,xmin=None,xmax=None,w=None,bksp
...
+        were rejected during the fit, if they are False. Data
+        rejected during the fit (if rejection is performed) are::
+
+            rejected = ingpm & np.invert(gpm)
+
+    yfit : `numpy.ndarray`_
+        The best-fitting model; shape is the same as ``xdata``.
+    reduced_chi : :obj:`float`
+        Reduced chi-square of the best-fitting model.
+    exit_status : :obj:`int`
```
Comment (jhennawi) [inline pypeit/utils.py]: Thanks for updating the docs.

File: pypeit/utils.py [hunk 11]
```diff
@@ -682,50 +713,58 @@ def bspline_profile(xdata, ydata, invvar, profile_basis, inmask = None, upper=5,
...
+    outmask = True if invvar.size == 1 else np.ones(invvar.shape, dtype=bool)
 
-    nin = np.sum(inmask)
-    msgs.info("Fitting npoly =" + "{:3d}".format(npoly) + " profile basis functions, nin=" + "{:3d}".format(nin) + " good pixels")
-    msgs.info("******************************  Iter  Chi^2  # rejected  Rel. fact   ******************************")
-    msgs.info("                              ----  -----  ----------  --------- ")
+    if ingpm is None:
+        ingpm = invvar > 0
 
+    if not quiet:
```
Comment (jhennawi) [inline pypeit/utils.py]: Thanks, we badly needed to suppress all this writing to the screen.

File: pypeit/utils.py [hunk 12]
```diff
@@ -682,50 +713,58 @@ def bspline_profile(xdata, ydata, invvar, profile_basis, inmask = None, upper=5,
...
     if not maskwork.any():
         msgs.error('No valid data points in bspline_profile!.')
-    else:
-        # Init bspline class
-        sset = pydl.bspline(xdata[maskwork], nord=nord, npoly=npoly, bkpt=bkpt, fullbkpt=fullbkpt,
-                                funcname='Bspline longslit special', **kwargs_bspline)
-        if maskwork.sum() < sset.nord:
+
+    # Init bspline class
+#    sset = pydl.bspline(xdata[maskwork], nord=nord, npoly=npoly, bkpt=bkpt, fullbkpt=fullbkpt,
```
Comment (jhennawi) [inline pypeit/utils.py]: Remove.

File: pypeit/utils.py [hunk 13]
```diff
@@ -682,50 +713,58 @@ def bspline_profile(xdata, ydata, invvar, profile_basis, inmask = None, upper=5,
...
+#                            funcname='Bspline longslit special', **kwargs_bspline)
+    sset = bspline.bspline(xdata[maskwork], nord=nord, npoly=npoly, bkpt=bkpt, fullbkpt=fullbkpt,
+                           funcname='Bspline longslit special', **kwargs_bspline)
+    if maskwork.sum() < sset.nord:
+        if not quiet:
             msgs.warn('Number of good data points fewer than nord.')
-            exit_status = 4
-            return sset, outmask, yfit, reduced_chi, exit_status
+        # TODO: Why isn't maskwork returned?
```
Comment (jhennawi) [inline pypeit/utils.py]: Because no rejection was performed, see below. Rejection is done with djs_reject.

File: pypeit/utils.py [hunk 14]
```diff
@@ -682,50 +713,58 @@ def bspline_profile(xdata, ydata, invvar, profile_basis, inmask = None, upper=5,
...
-    error = -1
-    qdone = False
+    error = -1                  # Indicates that the fit should be done
+    qdone = False               # True if rejection iterations are done
     exit_status = 0
     relative_factor = 1.0
-    tempin = np.copy(inmask)
-    while (error != 0 or qdone is False) and iiter <= maxiter and (exit_status == 0):
+    nrel = 0 if relative is None else len(relative)
+    # TODO: Why do we need both maskwork and tempin?
```
Comment (jhennawi) [inline pypeit/utils.py]: It has to do with how djs_reject works. I would need to think throught it carefully and go through that code.

File: pypeit/core/flux_calib.py [hunk 15]
```diff
@@ -1129,3 +1134,161 @@ def scale_in_filter(wave, flux, gpm, scale_dict):
 
     return scale
 
+
```
Comment (jhennawi) [inline pypeit/core/flux_calib.py]: I'm confused as to why this was added? Should this be moved to deprecated?

File: pypeit/core/pydl.py [hunk 16]
```diff
@@ -174,740 +174,793 @@ def djs_maskinterp(yval, mask, xval=None, axis=None, const=False):
...
-    kn = bw - 1
-    spot = np.arange(kn, dtype='i4') + 1
-    for j in range(n):
-        b[j] /= a[0, j]
-        b[j+spot] -= b[j]*a[spot, j]
-    spot = kn - np.arange(kn, dtype='i4')
-    for j in range(n-1, -1, -1):
-        b[j] = (b[j] - np.sum(a[spot, j] * b[j+spot]))/a[0, j]
-    return (-1, b)
+#class bspline(object):
```
[the remark this answers is not in the document]
Comment (kbwestfall) [reply pypeit/core/pydl.py]: Done.  Added `deprecated/pydl.py`.

File: pypeit/core/pydl.py [hunk 17]
```diff
@@ -174,740 +174,793 @@ def djs_maskinterp(yval, mask, xval=None, axis=None, const=False):
...
+#        b[j+spot] -= b[j]*a[spot,j]
+#
+#    spot = spot[::-1]
+#    for j in range(n-1, -1, -1):
+#        b[j] = (b[j] - np.sum(a[spot,j] * b[j+spot]))/a[0,j]
+#
+#    return -1, b
 
 
 def iterfit(xdata, ydata, invvar=None, inmask = None, upper=5, lower=5, x2=None,
```
[the remark this answers is not in the document]
Comment (kbwestfall) [reply pypeit/core/pydl.py]: Leaving it for now.

File: pypeit/core/pydl.py [hunk 18]
```diff
@@ -1055,210 +1106,270 @@ def iterfit(xdata, ydata, invvar=None, inmask = None, upper=5, lower=5, x2=None,
...
-    except AttributeError:
-        dt = np.float64
-    leg = np.ones((m, n), dtype=dt)
-    if m >= 2:
-        leg[1, :] = x
-    if m >= 3:
-        for k in range(2, m):
-            leg[k, :] = leg[k-1, :] * x
-    return leg
+## TODO: How important is it that the function return the last
```
[the remark this answers is not in the document]
Comment (kbwestfall) [reply pypeit/core/pydl.py]: Done.

File: pypeit/core/skysub.py [hunk 19]
```diff
@@ -293,16 +295,18 @@ def skyoptimal(wave, data, ivar, oprof, sortpix, sigrej=3.0, npoly=1, spatial=No
...
+                                        ingpm=mask1, fullbkpt=fullbkpt, upper=sigrej, lower=sigrej,
+                                        relative=relative,
+                                        kwargs_reject={'groupbadpix': True, 'maxrej': 1})
     else:
         msgs.warn('All pixels are masked in skyoptimal after first round of rejection. Not performing local sky subtraction.')
         return np.zeros_like(wave), np.zeros_like(wave), outmask
 
     ncoeff = npoly + nobj
-    skyset = pydl.bspline(None, fullbkpt=sset.breakpoints, nord=sset.nord, npoly=npoly)
+#    skyset = pydl.bspline(None, fullbkpt=sset.breakpoints, nord=sset.nord, npoly=npoly)
```
[the remark this answers is not in the document]
Comment (kbwestfall) [reply pypeit/core/skysub.py]: Done.

File: pypeit/core/skysub.py [hunk 20]
```diff
@@ -315,7 +319,8 @@ def skyoptimal(wave, data, ivar, oprof, sortpix, sigrej=3.0, npoly=1, spatial=No
     sky_bmodel, _ = skyset.value(wave, x2=spatial)
 
     obj_bmodel = np.zeros(sky_bmodel.shape)
-    objset = pydl.bspline(None, fullbkpt=sset.breakpoints, nord=sset.nord)
+#    objset = pydl.bspline(None, fullbkpt=sset.breakpoints, nord=sset.nord)
```
[the remark this answers is not in the document]
Comment (kbwestfall) [reply pypeit/core/skysub.py]: Done.

File: pypeit/core/skysub.py [hunk 21]
```diff
@@ -362,7 +367,8 @@ def optimal_bkpts(bkpts_optimal, bsp_min, piximg, sampmask, samp_frac=0.80,
     pix = pix[isrt]
     piximg_min = pix.min()
     piximg_max = pix.max()
-    bset0 = pydl.bspline(pix, nord=4, bkspace=bsp_min)
+#    bset0 = pydl.bspline(pix, nord=4, bkspace=bsp_min)
```
[the remark this answers is not in the document]
Comment (kbwestfall) [reply pypeit/core/skysub.py]: Done.

File: pypeit/flatfield.py (hunk 4 above)
Comment (kbwestfall) [reply pypeit/flatfield.py]: This was the original message.  Leaving it for now.

File: pypeit/flatfield.py (hunk 5 above)
Comment (kbwestfall) [reply pypeit/flatfield.py]: Done.

File: pypeit/flatfield.py (hunk 6 above)
Comment (kbwestfall) [reply pypeit/flatfield.py]: Done.

File: pypeit/reduce.py (hunk 7 above)
Comment (kbwestfall) [reply pypeit/reduce.py]: Done.

File: pypeit/spectrographs/keck_hires.py (hunk 8 above)
Comment (kbwestfall) [reply pypeit/spectrographs/keck_hires.py]: This was left over from the previous `pypeitimage_datamodel` branch.  I'll clean this up in `spec_datamodel`.

File: pypeit/utils.py (hunk 11 above)
Comment (kbwestfall) [reply pypeit/utils.py]: At the moment, it's still printed.  But we can change the default value of quiet to True.  Let's leave that for another PR, though.

File: pypeit/utils.py (hunk 12 above)
Comment (kbwestfall) [reply pypeit/utils.py]: Done.

File: pypeit/utils.py (hunk 13 above)
Comment (kbwestfall) [reply pypeit/utils.py]: Okay, thanks.

File: pypeit/utils.py (hunk 14 above)
Comment (kbwestfall) [reply pypeit/utils.py]: Okay.  In infinite time ...

**Pull request merged**